### ISSUE TITLE
[New Feature] [Polish] Show attachment previews for common file types

### ISSUE THREAD

**Issue by JmillsExpensify** [Expensify/App#12531]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Sign into any account on new.expensify.com.
2. Navigate to any DM.
2. Drag-and-drop or upload an HEIC file using the "+" to the left of the chat input. 

## Expected Result:
The HEIC file appear without a preview, even though it's a image format. 

![Screenshot 2022-11-03 at 14 48 35](https://user-images.githubusercontent.com/1127863/200406266-aa20cea6-8714-447b-a8f8-099d7468b7b2.png)

## Actual Result:
HEIC files should appear with a preview, like other images. In addition, we should minimally add previews for the following file types:

 - .xls / xlsx
 - .doc / docx
 - .ppt / pptx

## Workaround:
None

## Platform:
All platforms

- Web
- iOS
- Android
- Desktop App
- Mobile Web

**Version Number:*v1.2.23-7* 
**Reproducible in staging?:*Yes* 
**Reproducible in production?:*Yes* 
**Email or phone of affected tester (no customers):*jason@expensifail.com* 
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:*https://expensify.slack.com/archives/C01GTK53T8Q/p1667508523168689* Any additional supporting documentation
**Expensify/Expensify Issue URL:** 
**Issue reported by:*Expensify Internal* 
**Slack conversation:*https://expensify.slack.com/archives/C01GTK53T8Q/p1667508523168689* 

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by JmillsExpensify** [comment]
I'll actually take this one, as I'm working with @youssef-lr on it.

**Comment by JmillsExpensify** [comment]
Putting this issue on hold while we focus on bugs and related kinks in our initial implementation. There's no urgency on this one, so we can circle back on it at any time. Adding the `weekly` label in the meantime.

**Comment by JmillsExpensify** [comment]
Still not a priority and won't be for at least several weeks.

**Comment by JmillsExpensify** [comment]
Still not a priority and still won't be until WAQ is done.

**Comment by JmillsExpensify** [comment]
Still on hold for WAQ.

**Comment by JmillsExpensify** [comment]
@youssef-lr when you come back, I actually think that we should consider taking this one off hold.

**Comment by youssef-lr** [comment]
Currently focused on https://github.com/Expensify/Expensify/issues/242146, it's close to having PRs ready for review. I will start working on this immediately after I'm done.

**Comment by JmillsExpensify** [comment]
Nice, this is still a pretty nice polish item. We should probably do a pre-design on it in open-source before diving into any implementation.

**Comment by youssef-lr** [comment]
hey @JmillsExpensify, sorry I've been pretty swamped lately and didn't have time to get to this. I'm working on clearing some issues this week and start working on this next week.

**Comment by JmillsExpensify** [comment]
Nice, all good! I think it makes sense to take this through p/s and pre-design as well. We should also decide which comes first, as I think supporting all/more file types would come before this one. Also cc @marcaaron is there any overlap with link previews? Like does the service we're using do the same for uploaded file types?

**Comment by youssef-lr** [comment]
We're on track to make this open for contributors, we will mainly need to implement previewing of HEIC images if uploaded from desktop, and video files for starters. Is there anything else we should be previewing as well @JmillsExpensify?

**Comment by marcaaron** [comment]
@JmillsExpensify I think these problems are a bit different - internal attachments are not really "links". But maybe they can use the same UI component whenever it gets merged (development seems to have stalled there).

**Comment by JmillsExpensify** [comment]
Cool, makes sense. All in all, I still think we should pre-design this and also run it through a design doc before opening up for contributors.

**Comment by JmillsExpensify** [comment]
I still think this one is good to pick up and run a pre-design for. @youssef-lr is this on your list or should I look for other volunteers?

**Comment by youssef-lr** [comment]
I will post a predesign this week @JmillsExpensify!

**Comment by JmillsExpensify** [comment]
Not sure what the latest is on this one. Are you still working on it @youssef-lr ?

**Comment by youssef-lr** [comment]
I think we can make this available for contributors. We now support all file types so we can start working on showing previews for the common ones. What do you think?

**Comment by JmillsExpensify** [comment]
Sounds great, I'll float in Slack.

**Comment by marcaaron** [comment]
@wojtus7 who worked on Link Previews might be interested in this one as well?

**Comment by johnmlee101** [comment]
We should **not** display user-uploaded SVGs as they are a security concern.

**Comment by JmillsExpensify** [comment]
Updated the OP and removed SVG

**Comment by kowczarz** [comment]
Hello, I have 2 tasks (both are nearly finished) and I would like to start working on this issue by the end of the week.

**Comment by kowczarz** [comment]
Hi! In terms of HEIC image previews, on Android and iOS we can utilise [expo-image](https://docs.expo.dev/versions/latest/sdk/image/), however for browsers and electron (which is based on chromium browser) there is a problem, because [most of them don't support HEIC format](https://caniuse.com/?search=heic). Additionally, chrome is [not going to support the format in near future](https://bugs.chromium.org/p/chromium/issues/detail?id=843310), since they decided to focus on [AVIF](https://bugs.chromium.org/p/chromium/issues/detail?id=960620).

I did some research how others solve this issue:
- Slack - converts the image to .png, but keeps the original file and when the user downloads it, they receive the original .heic image.
- WhatsApp - converts the image to .jpeg and that’s how the images are being stored and downloaded
- Messenger - same as above
- Discord - same as WhatsApp

About other file types (.xls / xlsx, .doc / docx, .ppt / pptx) - the easiest way would be to create a preview on the backend and providing image/pdf so the app will be able to display it without additional calculations.

**Comment by mczernek** [comment]
> Hi! In terms of HEIC image previews, on Android and iOS we can utilise [expo-image](https://docs.expo.dev/versions/latest/sdk/image/), however for browsers and electron (which is based on chromium browser) there is a problem, because [most of them don't support HEIC format](https://caniuse.com/?search=heic). Additionally, chrome is [not going to support the format in near future](https://bugs.chromium.org/p/chromium/issues/detail?id=843310), since they decided to focus on [AVIF](https://bugs.chromium.org/p/chromium/issues/detail?id=960620).
> 
> I did some research how others solve this issue:
> 
> * Slack - converts the image to .png, but keeps the original file and when the user downloads it, they receive the original .heic image.
> * WhatsApp - converts the image to .jpeg and that’s how the images are being stored and downloaded
> * Messenger - same as above
> * Discord - same as WhatsApp
> 
> About other file types (.xls / xlsx, .doc / docx, .ppt / pptx) - the easiest way would be to create a preview on the backend and providing image/pdf so the app will be able to display it without additional calculations.

Adding `expo-image` opens door for other `expo` modules and I am all for it! We are actually considering proposing `expo-av` so it makes two potential use cases. There is some overhead with that, but I guess two use cases justify that. I also had some chat with expo folks and they are planning on reducing this overhead (meaning some additional packages being installed) in the near future.

**Comment by kowczarz** [comment]
During conversation I''ve just spotted that we can also consider enabling support to [AVIF](https://en.wikipedia.org/wiki/AVIF) files - it's an open source alternative to HEIC, which is used by some Android devices/cameras.

**Comment by kowczarz** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
We want to show previews of commonly used files:
- .xls / xlsx
- .doc / docx
- .ppt / pptx

### What is the root cause of that problem?
NewDot doesn't have an implementation of methods allowing users to preview mentioned file types.

### What changes do you think we should make in order to solve the problem?
Backend should generate and return pdf/jpeg with a preview of the file same as it's done currently in OldDot. Then we can utilize the current mechanism of previewing pdf/jpeg files.

**Comment by kowczarz** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
We want to display images in HEIC and AVIF formats in a similar way as we are displaying PNG and JPEG.

### What is the root cause of that problem?
None of non-beta browsers [support HEIC format](https://caniuse.com/?search=heic) and we don’t handle AVIF as an image file.

### What changes do you think we should make in order to solve the problem?
We need to enable converting any HEIC image into JPEG on the back-end. This will solve the issue of not displaying already sent attachmentsIn order to allow users preview images before uploading we could introduce `expo-image`. The library supports HEIC on mobile platforms and AVIF on all platforms. HEIC images are most common on mobile as a default format of photos.

Additional advantages of the `expo-image` are that it’s currently a state of the art in React-Native ecosystem, maintainers are responsive and well defined . Moreover, it works with the new architecture without downloading any additional code from github.

### Alternative solutions
- We can also store the original file and when the user clicks the `download` button, then we can provide the original image.
- We can just convert AVIF and HEIC images on the backend, and while uploading just display a generic placeholder.
- We can use a library that converts heic images to jpeg on the front-end, but this will highly affect the bundle size and will work slow on mobile devices.

**Comment by kowczarz** [comment]
@JmillsExpensify Do you have some thoughts on the proposals?

**Comment by kowczarz** [comment]
@youssef-lr did you manage to check the proposal?

**Comment by youssef-lr** [comment]
> We can use a library that converts heic images to jpeg on the front-end, but this will highly affect the bundle size and will work slow on mobile devices.

Ideally we'd want to convert HEIC in the frontend to have a preview ready for display immediately before uploading, it's essential to have the preview showing for the user before the click Send so the user can verify they're uploading the correct picture (and not something chosen by mistake)

**Comment by kowczarz** [comment]
@youssef-lr yes that would be the ideal approach, but there are limitations that raised some concerns about the [bundle size](https://expensify.slack.com/archives/C01GTK53T8Q/p1691059428599419?thread_ts=1691057390.605089&cid=C01GTK53T8Q) and overall performance of the conversion on mobile devices since the algorithm is implemented fully in JS without any native support.

**Comment by roryabraham** [comment]
Upon @JmillsExpensify's request, I spent a bit of time reviewing [this thread](https://expensify.slack.com/archives/C01GTK53T8Q/p1691057390605089). Here's a graphical summary of the decision tree before us as I see it:

```mermaid
flowchart LR
    SHOW_HEIC{"Show HEIC previews?"}
    CLIENT_CONVERT_HEIC("Convert HEIC on client.")
    DO_NOTHING("Do nothing.")
    PERFORMANCE_ACCEPTABLE{"Performance cost \nacceptable?"}
    YES_ALL(("yes on \nall platforms"))
    YES_WEB_ONLY(("yes, but \nonly on web"))
    NO_ALL(("no, not on \nany platform"))
    NO_BACKEND_CONVERT("Convert HEIC in the front-end,\nnever on the back-end")
    USE_EXPO("- Use expo-image\n- convert HEIC in front-end on web only\n- always convert HEIC in the back-end")

    SHOW_HEIC -- YES --> CLIENT_CONVERT_HEIC
    SHOW_HEIC -- NO --> DO_NOTHING

    CLIENT_CONVERT_HEIC --> PERFORMANCE_ACCEPTABLE
    PERFORMANCE_ACCEPTABLE --> YES_ALL
    PERFORMANCE_ACCEPTABLE --> YES_WEB_ONLY
    PERFORMANCE_ACCEPTABLE --> NO_ALL

    YES_ALL --> NO_BACKEND_CONVERT
    NO_ALL -- TOO BAD, NOT POSSIBLE --> SHOW_HEIC
    YES_WEB_ONLY --> USE_EXPO
```

**Comment by roryabraham** [comment]
Upon @JmillsExpensify's request, I spent a bit of time reviewing [this thread](https://expensify.slack.com/archives/C01GTK53T8Q/p1691057390605089). Here's a graphical summary of the decision tree before us as I see it:

```mermaid
flowchart LR
    SHOW_HEIC{"Show HEIC previews?"}
    YES(("Yes"))
    NO(("No"))
    CLIENT_CONVERT_HEIC("(2) Convert HEIC on client.")
    DO_NOTHING("(1)\nDo nothing.")
    PERFORMANCE_ACCEPTABLE{"Performance cost \nacceptable?"}
    FINE_FOR_UPLOADS(("it's good enough\n for previews \nin upload flow"))
    NOT_REALLY(("no, it's \na problem"))
    NOT_AT_ALL(("no way, it's \na blocker"))
    USE_EXPO("(3)\nUse expo-image, only \nconvert HEIC on web client")

    SHOW_HEIC --> NO -----> DO_NOTHING

    SHOW_HEIC --> YES --> CLIENT_CONVERT_HEIC
    CLIENT_CONVERT_HEIC --> PERFORMANCE_ACCEPTABLE

    PERFORMANCE_ACCEPTABLE --> FINE_FOR_UPLOADS
    PERFORMANCE_ACCEPTABLE --> NOT_REALLY
    PERFORMANCE_ACCEPTABLE --> NOT_AT_ALL

    NOT_REALLY --> USE_EXPO
    NOT_AT_ALL -- TOO BAD, NOT POSSIBLE --> SHOW_HEIC
```

**Comment by roryabraham** [comment]
Spent a while (probably longer than I should have) thinking about this, and my vote is that we just use `heic2any` for now to convert HEIC images in a cross-platform way: https://expensify.slack.com/archives/C01GTK53T8Q/p1694692977599789?thread_ts=1691057390.605089&cid=C01GTK53T8Q

**Comment by youssef-lr** [comment]
Latest update in this thread https://expensify.slack.com/archives/C01GTK53T8Q/p1691057390605089, it looks like we're going to give `heic2any` a shot.

**Comment by kowczarz** [comment]
I prepared a [POC](https://github.com/Expensify/App/pull/27933) and I ran some benchmarks, here the outcomes - with `heic2any` to convert a HEIC image taken with iPhone 13 pro to JPEG it takes about 3 seconds.
<img width="1446" alt="Screenshot 2023-09-21 at 12 37 42" src="https://github.com/Expensify/App/assets/91068263/6f9f314e-f891-4f2c-bc22-737aac9a7be0">

**Comment by roryabraham** [comment]
3 seconds seems a touch slow but within an acceptable range for a user to wait for the preview to appear. People are pretty used to waiting a moment for an image preview to appear.

**Comment by kowczarz** [comment]
@roryabraham to improve the experience we can consider uploading the image in the meantime, so we will skip upload time after sending the image.

**Comment by kowczarz** [comment]
FYI I will be OOO for the whole following week.

**Comment by JmillsExpensify** [comment]
Agree that 3 seconds feels pretty long, though I think it's fine to start there.

**Comment by youssef-lr** [comment]
Not overdue.

**Comment by kowczarz** [comment]
I'm back from OOO and continued the discussion via slack

**Comment by youssef-lr** [comment]
Not overdue.

**Comment by JmillsExpensify** [comment]
Blocked on backend changes by an internal Expensify engineer.

**Comment by youssef-lr** [comment]
@JmillsExpensify sorry, what backend changes are needed for this?

**Comment by youssef-lr** [comment]
Not overdue.

**Comment by JmillsExpensify** [comment]
@kowczarz Can you clarify @youssef-lr's question above?

**Comment by JmillsExpensify** [comment]
Also for visibility, there is a long [thread in Slack](https://expensify.slack.com/archives/C01GTK53T8Q/p1691057390605089) discussing this topic.

**Comment by JmillsExpensify** [comment]
Still an issue we should tackle since it's coming up in user testing.

**Comment by JmillsExpensify** [comment]
Same

**Comment by youssef-lr** [comment]
@kowczarz now that we have started using `expo-image`, do you think you can resume your PR using it instead?

**Comment by kowczarz** [comment]
Hi @youssef-lr currently I'm focused on Video Playback feature right now, but @BartoszGrajdek will take this issue and he can resume the PR.

**Comment by BartoszGrajdek** [comment]
Just adding a link to current discussion from Slack: https://expensify.slack.com/archives/C01GTK53T8Q/p1705930052644969

**Comment by BartoszGrajdek** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
We want to display images in HEIC formats (default iOS camera photo format) in a similar way as we are displaying PNG and JPEG.

### What is the root cause of that problem?
Aside from Safari no browsers [support HEIC format](https://caniuse.com/?search=heic), we can display HEIC images on iOS, but there are also some problems on Android.

### What changes do you think we should make in order to solve the problem?

Since we have `expo-image` now available in the project we should utilize it and make sure that the images on mobile are handled correctly. 

Now the problem is that as per [docs](https://docs.expo.dev/versions/latest/sdk/image/) `expo-image` doesn't currently provide support for HEIC files on the web. The solution is to add `heic2any` library on web/desktop ([here's the request](https://github.com/Expensify/App/issues/35798) to add it into Expensify), that way all the devices will be able to handle HEIC images.

### What alternative solutions did you explore? (Optional)

**Reminder:** Please use plain English, be brief and avoid jargon. Feel free to use images, charts or pseudo-code if necessary. Do not post large multi-line diffs or write walls of text. Do not create PRs unless you have been hired for this job.

I see two alternatives:
1. We could generate JPEG files on the backend, and only provide the original files when the user wants to download the file. The problem here is that on web/desktop whenever a user tries to attach a HEIC file they won't be able to see the preview, only once we get the preview generated will it be shown and this also requires us to store 2 files for each HEIC image.
2. We could ignore web/desktop previews altogether and only display them on mobile (just for HEIC files) - though I'm not a fan of that option.

**Comment by JmillsExpensify** [comment]
I think we agreed to table this one, right? Going to close based on the Slack discussion, though please re-open if you disagree.
